충남대 FE 정지원 - 3단계 테마 상품 목록 페이지 구현하기

jjw5655 opened the pull request:
박진영 멘토님 안녕하세요! 벌써 마지막 단계네요.

이전 단계에서 리뷰 남겨주신 내용을 모두 반영했습니다.
구현 사항은 readme와 커밋메시지 참고 부탁드립니다. 감사합니다!

jjw5655 committed:
refactor: 공용 apiClient 인스턴스 사용으로 axios 호출 방식 개선

- loginApi, postOrder 함수에서 axios.post와 baseURL 직접 사용 대신 apiClient.post와 상대 경로 사용
- 중복된 baseURL 지정 제거로 코드 간결화 및 유지보수성 향상

jjw5655 committed:
refactor: 주문 처리 핸들러 역할 분리 및 타입 명확화

- 주문 데이터 가공, 안내 메시지 생성, 수량/가격 계산, 에러 핸들링을 별도 함수로 분리하여 가독성 및 유지보수성 향상
- product 관련 파라미터 타입을 ProductSummary로 명확히 지정
- 불필요한 Product 타입 import 제거

jjw5655 committed:
feat: 테마 상품 목록 페이지 라우트 및 기본 구조 추가

- /themes/:themeId 경로에 ThemeProductListPage 라우트 추가
- ThemeProductListPage 컴포넌트 기본 파일 생성
- 라우트 상수(ROUTE_THEME_PRODUCTS) 정의 및 적용
- pages/index.ts에 ThemeProductListPage export 추가

jjw5655 committed:
feat: 테마 정보(히어로 영역) API 연동 및 스타일 설정

- /api/themes/:themeId/info API 연동을 위한 getThemeInfo 함수 추가
- ThemeProductListPage에서 themeId로 테마 정보 fetch 및 히어로 영역 표시
- 404 에러 발생 시 홈으로 리다이렉트
- Theme 타입 확장 및 히어로 영역 styled-components 적용
- 로딩, 에러, 정상 상태 UI 처리

jjw5655 committed:
refactor: 상단바 타이틀을 항상 '선물하기'로 고정

jjw5655 committed:
feat: 카테고리 클릭 시 테마 상품 목록 페이지로 라우팅

jjw5655 committed:
feat: 테마 상품 리스트 API 연동 및 기본 렌더링

- /api/themes/:themeId/products API 연동(getThemeProducts) 및 실제 응답 구조에 맞게 파싱
- ThemeProductListPage에서 상품 리스트 fetch 및 ProductGrid로 기본 렌더링
- 상품이 없을 때 빈 상태 UI 표시
- 상품 클릭 시 주문 페이지(/order/:productId)로 이동 기능 추가

jjw5655 committed:
feat: Intersection Observer를 활용한 무한 스크롤 구현

- useInfiniteScroll 커스텀 훅 구현 및 적용
- ThemeProductListPage에서 상품 하단 도달 시 추가 상품 자동 로드
- getThemeProducts에 cursor 파라미터 추가로 페이지네이션 지원
- 로딩/마지막 상품 안내 UI 추가
- ProductGrid 상품 개수 제한 해제, 무한 스크롤 시 전체 상품 표시

jjw5655 committed:
docs: 미션 3 - 3단계 과제 요구사항으로 README.md 업데이트

ed-jinyoung-park commented on src/pages/OrderPage.tsx:
함수단위로 리팩토링을 잘 반영해주셨네요 👍

ed-jinyoung-park commented on src/pages/ThemeProductListPage.tsx:
```diff
@@ -0,0 +1,138 @@
+        } else {
+          setError('테마 정보를 불러오지 못했습니다.');
+        }
+      })
```
Promise.all을 사용하신 이유가 있을까요? 
Promise.all은 비동기 작업 여러개가 모두 수행된 결과가 필요할 경우에 사용하는데요.

지금 상황에서는 `getThemeInfo`와 `getThemeProducts`는 독립적으로 각각 실행되고 둘 중 아무거나 먼저 실행되면 그 결과를 받아 함수에 그리도록 병렬처리를 하는게 적절해보여요. Promise.all을 사용하면 직렬처리를 하는 꼴이 되어 먼저 그려질 수 있는 UI를 그리지 못하게 됩니다. 


Promise.all 사용시 
```
getThemeInfo ......완료 ........(getThemeProducts를 기다림)..........setTheme
getThemProducts ........................................완료....setProducts
```

개별 호출시
```
getThemeInfo ......완료.setTheme(theme화면 미리 그릴수 있음)
getThemProducts .................완료...setProducts
```

또한 error catch 상황에서도 어떤 api 호출에 대한 error인지 코드상으로 판단하기 어렵습니다. 

아래 두가지를 액션아이템으로 참고해서 수정해보세요!

1) Promise.all 대신 개별 호출로 수정
2) getThemeInfo, getThemeProducts를 useFetch hook을 활용하려 리팩토링 (status, error, data라는 상태를 일관적으로 관리 가능)

jjw5655 replied:
사용자가 해당 페이지에 처음 들어간 경우를 생각해봤을때, 하나가 먼저 뜨고 나머지 하나가 뜨는 것보다 "로딩중" 상태가 뜨다가 2개가 동시에 뜨는게 더 낫지않을까? 싶어서 promise.all로 작성했던 것 같습니다. 한 섹션만 먼저 뜨고 나머지가 뜨게되면 경우에 따라 레이아웃이 갑자기 바뀌거나, 더 버벅인다고 느낄수도 있을 것 같아서요.

찾아보니 여러 대형 서비스들은 독립적으로 그리는 UX를 더 많이 사용하고있다고 하네요. 멘토님 조언대로 개별호출+useFetch활용으로 수정했습니다. 말씀해주신 error catch를 고려해봐도 분리하는게 더 나아보입니다.

레이아웃이 갑자기 바뀌게 될 수도 있다는 문제점은 로딩중에 스켈레톤을 보여주는 식으로 개선할 수 있을 것 같아요.

ed-jinyoung-park replied:
> 사용자가 해당 페이지에 처음 들어간 경우를 생각해봤을때, 하나가 먼저 뜨고 나머지 하나가 뜨는 것보다 "로딩중" 상태가 뜨다가 2개가 동시에 뜨는게 더 낫지않을까? 싶어서 promise.all로 작성했던 것 같습니다. 한 섹션만 먼저 뜨고 나머지가 뜨게되면 경우에 따라 레이아웃이 갑자기 바뀌거나, 더 버벅인다고 느낄수도 있을 것 같아서요.

이런 이유가 있으셨군요. 어플리케이션 특성에 따라 지원님이 말씀하신 것처럼 한꺼번에 레이아웃을 보여줘야 되는 경우도 있으니 케바케로 생각해주시면 될 것 같아요! 

그렇지만 말씀해주신대로 대부분 잘 만들어진 어플리케이션에서는 API가 필요한 섹션별로 스켈레톤을 먼저 보여주고, API 호출이 끝난 섹션을 먼저 보여주는 UX를 보여주곤 합니다. 이번 프로젝트는 그런 UX가 적절하고 API별로 에러를 별도로 다뤄줘야 하기 때문에 위와 같은 의견을 드렸습니다!

ed-jinyoung-park reviewed (changes requested):
지원님 고생하셨습니다!

다른 요구사항은 잘 구현해주셨는데요. 

무한스크롤에서 마지막 아이템을 끝까지 스크롤했을때 추가 아이템이 불러와져서 다소 부자연스러운 면이 있는 것 같습니다. 
아마 비어있는 부분까지 grid로 높이가 잡혀있어서 무한스크롤 관찰용 div 감지가 늦게 되는것 같은데 css style 수정으로 해결해보실 수 있을 것 같아요. 추가로 Promise.all 관련 코멘트도 남겼으니 1) 무한스크롤, 2) Promise.all 관련 두가지 이슈 수정해주시면 다시 확인해보겠습니다!

(무한스크롤 gif 첨부드립니다)
![무한스크롤](https://github.com/user-attachments/assets/f5ff59fa-09b4-4add-ad75-b5941a532f7e)

jjw5655 committed:
refactor: Promise.all 제거 및 API 에러 처리 방식 통일

- getThemeInfo, getThemeProducts를 Promise.all 대신 각각 개별적으로 호출하도록 변경
- 각 API의 에러 처리에서 404 발생 시 홈으로 이동, 그 외에는 setError 및 console.error로 에러 메시지와 로그를 남기도록 통일
- 코드 가독성과 유지보수성을 위해 에러 처리 방식 일관성 강화

jjw5655 committed:
fix: ProductGrid의 grid row 동적 처리로 무한스크롤 공백 문제 해결

- grid-template-rows를 제거하고 grid-auto-rows: 1fr로 변경
- 상품 개수만큼만 grid row가 생성되어, 불필요한 공백 없이 무한스크롤이 자연스럽게 동작하도록 개선

jjw5655 commented:
무한스크롤 문제도 해결했습니다. 기존의 랭킹섹션에서 사용하던 컴포넌트의 css를 그대로 사용하니 생긴 문제였습니다. grid-auto-rows: 1fr로 변경하였습니다.

gif까지 첨부까지.. 감사드립니다!

ed-jinyoung-park commented on src/pages/ThemeProductListPage.tsx:
```diff
@@ -0,0 +1,154 @@
+        if (err.response && err.response.status === 404) {
+          navigate(ROUTE_HOME, { replace: true });
+        } else {
+          setError('테마 정보를 불러오지 못했습니다.');
```
error 상태가 getThemeInfo, getThemeProducts 둘다 공용으로 사용하는 경우라면 둘다 에러가 나서 에러 메시지를 2개 보여줘야 할 때 이슈가 있을 것 같아요. api별로 가져오는 데이터에 대한 상태를 별도로 정의한 것처럼 에러 상태도 별도로 관리해주는게 에러 관리 및 유지보수 측면에서 적절할 것 같습니다.
api별로 데이터, 로딩중 상태, 에러 상태가 공통으로 있으니 useFetch hook으로 아래와 같이 인터페이스를 통일해보는 것도 좋을 것 같습니다!

```ts
// 예시코드
  const {data: themeInfo, isLoading: isThemeInfoLoading, error: isThemeInfoError}  = useFetch(getThemeInfo)
  const {data: themeproducts, isLoading: isThemeProductsLoading, error: isThemeProductsError}  = useFetch(getThemeProducts)

```

ed-jinyoung-park reviewed (approved):
피드백 잘 반영해주셨네요 👍 

좀 더 리팩토링해볼수 있는 부분을 코멘트로 남겼으니 확인해주세요!

3번째 미션 진행하시느라 고생많으셨습니다 마지막 미션도 응원할게요 화이팅입니다~
